`WCLocationModel` to `Room`

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: [AINFRA-553](https://linear.app/a8c/issue/AINFRA-553/migrate-wclocations)

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCLocationModel` from WellSql to Room.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

1. Open the mobile app
2. Create new order
3. Tap on `Add customer details`
4. Tap on `+` to add a new customer
5. See that `Country` dropdown has countries and when you select a country with regions (e.g. Japan), you can select a region/state.


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="1062" alt="image" src="https://github.com/user-attachments/assets/f519ad5d-9387-44b6-88ad-151499b3ba03" />


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Migrate `WCLocationModel` to Room

wzieba committed:
Merge branch 'trunk' into migrate_location

wzieba committed:
Adjust unit tests to the new `WCLocationModel`

Return empty list if `country` is empty for `getStates` method. Aligns with previous implementation.

wzieba committed:
Make `WCCountryMapper` an object

This is a tiny, static utility: there's no need to make it an object. This way, it won't increase the size of DI graph and not allow to mock it in tests.

wzieba committed:
Improve readability of `WCDataStoreTest`

Use plain assertions (works now as `WCLocationModel` is immutable data class now). Update names of tests

wzieba committed:
Drop `WCLocationModel` table from WellSql database

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCDataStore.kt:
```diff
@@ -8,29 +9,32 @@ import org.wordpress.android.fluxc.network.rest.wpcom.wc.WooError
      */
-    fun getStates(country: String): List<WCLocationModel> =
-            WCCountriesSqlUtils.getStates(country)
+    fun getStates(country: String): List<WCLocationModel> = if (country.isEmpty()) {
+        emptyList()
+    } else {
+        runBlocking { locationsDao.getStates(country) }
```
Nope, it involves changes in too many places of the codebase.

ParaskP7 commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/WellSqlConfig.kt:
```diff
@@ -2194,6 +2194,10 @@ open class WellSqlConfig : DefaultWellConfig {
                 }
+
+                217 -> migrateAddOn(ADDON_WOOCOMMERCE, version) {
+                    db.execSQL("DROP TABLE IF EXISTS WCLocationModel")
```
Warn (⚠️): As I was testing this, before me starting with the actual review, I initially thought that we forgot to drop the table from `wp-fluxc`. While reviewing the change I now understand that the query is actually wrong, we need to use `WCLocations` instead, using table name, not the model name. 😊

wzieba replied:
Ah true. I should've checked this, even by searching the name in this file (and find `CREATE TABLE` part). Addressed in 8f98a1b179218b8d4f4395c33a13c284dcf0eda4

ParaskP7 replied:
Awesome, thanks for addressing this @wzieba ! ❤️ 🚀

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/data/WCCountryMapper.kt:
```diff
@@ -1,20 +1,19 @@
 
-class WCCountryMapper
-@Inject constructor() {
+object WCCountryMapper {
```
Praise (❤️): Good call converting this to an object. Also, you mention that doing so it will not allow to mock it in tests, interesting... 😄

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/data/WCLocationModel.kt:
```diff
@@ -1,20 +1,13 @@
+
+@Entity(
+    tableName = "LocationEntity",
+    primaryKeys = ["parentCode", "code", "name"]
```
Question (❓): I wonder why you chose to add the `name` into the primary key combination, is it really needed? 🤔 

FYI: I then also noticed our friend the Copilot [mentioning](https://github.com/woocommerce/woocommerce-android/pull/14219#pullrequestreview-2943452009) a similar concern.

> Including name in the primary key is unusual since names can change; this may lead to duplicate rows if a name is updated. Consider using only parentCode and code, or introduce a synthetic ID.
> 
> ```
> primaryKeys = ["parentCode", "code", "name"]
> ```

wzieba replied:
Good point, thanks 👍 I agree, (`parentCode` x `code`) should be te primary key: af4442ec9a5fa1498dd355747d11b1ef6198261a

ParaskP7 replied:
Awesome, thanks for the change @wzieba ! 🚀 🙇

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCDataStore.kt:
```diff
@@ -8,29 +9,32 @@ import org.wordpress.android.fluxc.network.rest.wpcom.wc.WooError
-    fun getStates(country: String): List<WCLocationModel> =
-            WCCountriesSqlUtils.getStates(country)
+    fun getStates(country: String): List<WCLocationModel> = if (country.isEmpty()) {
+        emptyList()
```
Praise (❤️): Nicely done with adding that back, I almost missed this, or would have missed this during review! 🥇

ParaskP7 reviewed (changes requested):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected (ish), good job on yet another migration! 🌟 

-----

I have left a warning (⚠️) and a question (❓) comment for you. I am **NOT** going to approve this, requesting changes instead, just because of this single warning (⚠️) comment. 🙏

wzieba committed:
Fix incorrect name of WellSql "locations" table

wzieba committed:
Remove `name` from primary key of `WCLocationModel`

"Including name in the primary key is unusual since names can change; this may lead to duplicate rows if a name is updated. Consider using only parentCode and code, or introduce a synthetic ID."
